composer/callbacks/checkpoint_saver.py

```diff
@@ -411,14 +415,28 @@ def _save_checkpoint(self, state: State, logger: Logger):
                 remote_prefix = state.sharded_ckpt_prefix_dir
                 remote_file_name = os.path.join(
                     pathlib.Path(remote_file_name).parent, remote_prefix,
-                    pathlib.Path(remote_file_name).name)
+                    _TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME)
                 remote_file_name = format_name_with_dist_and_time(remote_file_name, state.run_name, state.timestamp)
+
+                # Upload metadata file.
+                if dist.get_global_rank() == 0:
+                    metadata_local_file_name = os.path.join(
+                        Path(saved_path).parent,
+                        _TORCH_DISTRIBUTED_CHECKPOINTS_METADATA_FILENAME
+                        )
+                    metadata_local_file_name = format_name_with_dist_and_time(metadata_local_file_name, state.run_name, state.timestamp)
+                    metadata_remote_file_name = os.path.join(
+                        Path(remote_file_name).parent,
+                        _TORCH_DISTRIBUTED_CHECKPOINTS_METADATA_FILENAME)
+                    metadata_remote_file_name = format_name_with_dist_and_time(metadata_remote_file_name, state.run_name, state.timestamp)
```

Comment:
It's a little weird that the intermediate values have the same name as the final value... can we rename that?

Also, can you put a comment on what metadata actually is?

Reply:
done